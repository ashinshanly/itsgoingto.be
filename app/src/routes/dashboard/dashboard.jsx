import React from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import { browserHistory } from 'react-router'
import { hasUserSelector } from 'store/user/selectors'
import { setLoading } from 'store/loader/actions'
import PollTable from './components/poll-table'
import Stats from './components/stats'
import './dashboard.scss'

class Dashboard extends React.Component {
  componentDidMount = () => this.checkPermissions()

  componentDidUpdate = () => this.checkPermissions()

  checkPermissions = () => {
    const { hasUser } = this.props

    if (!hasUser) {
      this.props.setLoading(true)
      browserHistory.push('/login')
    } else {
      this.props.setLoading(false)
    }
  }

  render () {
    const { hasUser } = this.props

    if (hasUser) {
      return (
        <div>
          <div className='container info-container'>
            <Stats />
          </div>
          <div className='container panel-container'>
            <PollTable />
          </div>
        </div>
      )
    }

    return <div />
  }
}

Dashboard.propTypes = {
  hasUser    : PropTypes.bool.isRequired,
  setLoading : PropTypes.func.isRequired,
}

const mapStateToProps = (state) => ({
  hasUser : hasUserSelector(state),
})

const mapDispatchToProps = (dispatch) => ({
  setLoading : (value) => dispatch(setLoading(value)),
})

export default connect(mapStateToProps, mapDispatchToProps)(Dashboard)
