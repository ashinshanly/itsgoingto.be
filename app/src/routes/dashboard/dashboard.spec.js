/* global expect, jest */
import React from 'react'
import { shallow } from 'enzyme'
import { browserHistory } from 'react-router'
import { Dashboard } from './dashboard'

browserHistory.push = jest.fn()

const props = {
  hasUser: false,
  setLoading: jest.fn(),
}

describe('(Route) Login', () => {
  describe('(Lifecycle) componentDidMount', () => {
    describe('checkPermissions', () => {
      beforeEach(() => {
        shallow(<Dashboard {...props} />)
      })

      afterEach(() => {
        jest.clearAllMocks()
      })

      describe('hasUser false', () => {
        it('should call setLoading true', () => {
          expect(props.setLoading).toHaveBeenCalledTimes(1)
          expect(props.setLoading).toHaveBeenCalledWith(true)
        })

        it('should redirect to /login', () => {
          expect(browserHistory.push).toHaveBeenCalledTimes(1)
          expect(browserHistory.push).toHaveBeenCalledWith('/login')
        })
      })

      describe('hasUser true', () => {
        it('should call setLoading false', () => {
          jest.clearAllMocks()
          shallow(<Dashboard {...props} hasUser />)

          expect(props.setLoading).toHaveBeenCalledTimes(1)
          expect(props.setLoading).toHaveBeenCalledWith(false)
        })
      })
    })
  })

  describe('(Lifecycle) componentDidUpdate', () => {
    describe('checkPermissions', () => {
      beforeEach(() => {
        const wrapper = shallow(<Dashboard {...props} />)
        jest.clearAllMocks()
        wrapper.setProps()
      })

      afterEach(() => {
        jest.clearAllMocks()
      })

      describe('hasUser false', () => {
        it('should call setLoading true', () => {
          expect(props.setLoading).toHaveBeenCalledTimes(1)
          expect(props.setLoading).toHaveBeenCalledWith(true)
        })

        it('should redirect to /login', () => {
          expect(browserHistory.push).toHaveBeenCalledTimes(1)
          expect(browserHistory.push).toHaveBeenCalledWith('/login')
        })
      })

      describe('hasUser true', () => {
        it('should call setLoading false', () => {
          const wrapper = shallow(<Dashboard {...props} hasUser />)
          jest.clearAllMocks()
          wrapper.setProps()

          expect(props.setLoading).toHaveBeenCalledTimes(1)
          expect(props.setLoading).toHaveBeenCalledWith(false)
        })
      })
    })
  })

  describe('(Render)', () => {
    let wrapper

    beforeEach(() => {
      wrapper = shallow(<Dashboard {...props} />)
    })

    afterEach(() => {
      jest.clearAllMocks()
    })

    describe('hasUser false', () => {
      it('matches snapshot', () => {
        expect(wrapper).toMatchSnapshot()
      })
    })

    describe('hasUser true', () => {
      it('matches snapshot', () => {
        wrapper.setProps({ hasUser: true })
        expect(wrapper).toMatchSnapshot()
      })
    })
  })
})
