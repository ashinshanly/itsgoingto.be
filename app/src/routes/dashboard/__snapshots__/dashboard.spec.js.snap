// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`(Route) Login (Render) hasUser false matches snapshot 1`] = `<div />`;

exports[`(Route) Login (Render) hasUser true matches snapshot 1`] = `
<div>
  <div
    className="container info-container"
  >
    <Connect(Stats) />
  </div>
  <div
    className="container panel-container"
  >
    <Connect(PollTable) />
  </div>
</div>
`;
